src/main/java/com/minecolonies/coremod/entity/ai/citizen/miner/EntityAIQuarrier.java

```diff
@@ -0,0 +1,725 @@
+    }
+
+    @Override
+    public int getPlaceSpeedLevel()
+    {
+        return getSecondarySkillLevel();
+    }
+
+    @Override
+    protected int getActionsDoneUntilDumping()
+    {
+        return getOwnBuilding().getBuildingLevel() * MAX_BLOCKS_MINED;
+    }
+
+    @Override
+    protected void updateRenderMetaData()
+    {
+        worker.setRenderMetadata(getRenderMetaStone() + getRenderMetaTorch());
+    }
+
+    /**
+     * Get render data to render torches at the backpack if in inventory.
+     *
+     * @return metaData String if so.
+     */
+    @NotNull
+    private String getRenderMetaTorch()
+    {
+        if (worker.getCitizenInventoryHandler().hasItemInInventory(Items.TORCH))
+        {
+            return RENDER_META_TORCH;
+        }
+        return "";
+    }
+
+    /**
+     * Get render data to render stone in the backpack if cobble in inventory.
+     *
+     * @return metaData String if so.
+     */
+    @NotNull
+    private String getRenderMetaStone()
+    {
+        if (worker.getCitizenInventoryHandler().hasItemInInventory(getMainFillBlock()))
+        {
+            return RENDER_META_STONE;
+        }
+        return "";
+    }
+
+    @Override
+    public IAIState doMining()
+    {
+        if (blockToMine == null)
+        {
+            return BUILDING_STEP;
+        }
+
+        for (final Direction direction : Direction.values())
+        {
+            final BlockPos pos = blockToMine.relative(direction);
+            final BlockState surroundingState = world.getBlockState(pos);
+
+            final FluidState fluid = world.getFluidState(pos);
+            if (surroundingState.getBlock() == Blocks.LAVA || (fluid != null && !fluid.isEmpty() && (fluid.getType() == Fluids.LAVA || fluid.getType() == Fluids.FLOWING_LAVA)) || SurfaceType.isWater(world, pos, surroundingState, fluid))
+            {
+                setBlockFromInventory(pos, getMainFillBlock());
+            }
+        }
+
+        final BlockState blockState = world.getBlockState(blockToMine);
+        if (!IColonyManager.getInstance().getCompatibilityManager().isOre(blockState))
+        {
+            blockToMine = getSurroundingOreOrDefault(blockToMine);
+        }
+
+        if (world.getBlockState(blockToMine).getBlock() instanceof AirBlock)
+        {
+            blockToMine = null;
+            return BUILDING_STEP;
+        }
+
+        if (!mineBlock(blockToMine, getCurrentWorkingPosition()))
+        {
+            worker.swing(Hand.MAIN_HAND);
+            return getState();
+        }
+
+        blockToMine = getSurroundingOreOrDefault(blockToMine);
+        if (IColonyManager.getInstance().getCompatibilityManager().isOre(world.getBlockState(blockToMine)))
+        {
+            return getState();
+        }
+
+        worker.decreaseSaturationForContinuousAction();
+        blockToMine  = null;
+        return BUILDING_STEP;
+    }
+
+    private BlockPos getSurroundingOreOrDefault(final BlockPos pos)
```

Comment:
Don't think this is needed for the quarry? could be one of the things which makes the difference between miner and quarry

Reply:
Good idea